ECHO_modules/utilities.py

```diff
@@ -376,10 +379,111 @@ def get_active_facilities( state, region_type, regions_selected ):
             # Get only those in the selected counties.
             df_active = get_facs_in_counties(df_active, regions_selected)
     except pd.errors.EmptyDataError:
-            df_active = None
+        df_active = None
 
     return df_active
 
+def aggregate_by_facility(records, program, other_records = False):
+  '''
+  Aggregate a set of records by facility IDs, using sum or count operations. 
+  Enables point symbol mapping. 
+  Other facilities in the selection (e.g. facilities in Snohomish County *without* 
+  reported CWA violations) can be identified and retrieved when the diff flag is True
+
+  Parameters
+  ----------
+  records : DataSetResults object
+      The records to aggregate. records should be a DataSetResults object created from
+      a database query. In the :
+      ds = make_data_sets(["CWA Violations"]) # Create a DataSet for handling the data
+      snohomish_cwa_violations = ds["CWA Violations"].store_results(region_type="County", region_value=("SNOHOMISH",) state="WA") # Store results for this DataSet as a DataSetResults object
+  program : String
+      The name of the program, usually available from records.dataset.name
+  other_records : Boolean
+      When True, will retrieve other facilities in the selection 
+      (e.g. facilities in Snohomish County *without* reported CWA violations)
+  
+  Returns
+  -------
+  A dictionary containing:
+    the aggregated results
+    active facilities regulated under this program, but without recorded violations, inspections, or whatever the metric is (e.g. violations)
+    the name of the new field that counts or sums up the relevant metric (e.g. violations) 
+  '''
+
+  data = records.dataframe
+  diff = None
+
+  def differ(input, program):
+    '''
+    Helper function to sort facilities in this program (input) from the full list of faciliities regulated under the program (active)
+    '''
+    active = get_active_facilities(records.state, records.region_type, records.region_value )
+
+    diff = list(
+        set(active[records.dataset.echo_type + "_IDS"]) - set(input[records.dataset.idx_field])
+        ) 
+    
+    # get rid of NaNs - probably no program IDs
+    diff = [x for x in diff if str(x) != 'nan']
+    
+    # ^ Not perfect given that some facilities have multiple NPDES_IDs
+    # Below return the full ECHO_EXPORTER details for facilities without violations, penalties, or inspections
+    diff = active.loc[active[records.dataset.echo_type + "_IDS"].isin(diff)] 
+    return diff
```

Comment:
This isn't a robust way of differentiating facilities because of the way EPA stores program IDs.

The differ function is meant to take data like:
all_facilities = [A, B, C, D, E]
facilities_with_inspections = [C, D]
and calculate:
facilities_without_inspections = [A, B, E]

But in reality, facility/program IDs are more like:
all_facilities = [A X, B Y, C, D Z, E]
facilities_with_inspections = [C, D]
So the resulting list of facilities without inspections would incorrectly be:
A X, B Y, D Z
even though D does have an inspection.

Just need a way to parse apart program IDs, ideally without having to call up the database to look at the EXP_PGM table.